src/graphics/vulkan.zig

```diff
@@ -0,0 +1,150 @@
+	VK_ERROR_INCOMPATIBLE_DRIVER = -9,
+	VK_ERROR_TOO_MANY_OBJECTS = -10,
+	VK_ERROR_FORMAT_NOT_SUPPORTED = -11,
+	VK_ERROR_FRAGMENTED_POOL = -12,
+	VK_ERROR_UNKNOWN = -13,
+	VK_ERROR_OUT_OF_POOL_MEMORY = -1000069000,
+	VK_ERROR_INVALID_EXTERNAL_HANDLE = -1000072003,
+	VK_ERROR_FRAGMENTATION = -1000161000,
+	VK_ERROR_INVALID_OPAQUE_CAPTURE_ADDRESS = -1000257000,
+	VK_PIPELINE_COMPILE_REQUIRED = 1000297000,
+	VK_ERROR_NOT_PERMITTED = -1000174001,
+	VK_ERROR_SURFACE_LOST_KHR = -1000000000,
+	VK_ERROR_NATIVE_WINDOW_IN_USE_KHR = -1000000001,
+	VK_SUBOPTIMAL_KHR = 1000001003,
+	VK_ERROR_OUT_OF_DATE_KHR = -1000001004,
+	VK_ERROR_INCOMPATIBLE_DISPLAY_KHR = -1000003001,
+	VK_ERROR_VALIDATION_FAILED_EXT = -1000011001,
+	VK_ERROR_INVALID_SHADER_NV = -1000012000,
+	VK_ERROR_IMAGE_USAGE_NOT_SUPPORTED_KHR = -1000023000,
+	VK_ERROR_VIDEO_PICTURE_LAYOUT_NOT_SUPPORTED_KHR = -1000023001,
+	VK_ERROR_VIDEO_PROFILE_OPERATION_NOT_SUPPORTED_KHR = -1000023002,
+	VK_ERROR_VIDEO_PROFILE_FORMAT_NOT_SUPPORTED_KHR = -1000023003,
+	VK_ERROR_VIDEO_PROFILE_CODEC_NOT_SUPPORTED_KHR = -1000023004,
+	VK_ERROR_VIDEO_STD_VERSION_NOT_SUPPORTED_KHR = -1000023005,
+	VK_ERROR_INVALID_DRM_FORMAT_MODIFIER_PLANE_LAYOUT_EXT = -1000158000,
+	VK_ERROR_FULL_SCREEN_EXCLUSIVE_MODE_LOST_EXT = -1000255000,
+	VK_THREAD_IDLE_KHR = 1000268000,
+	VK_THREAD_DONE_KHR = 1000268001,
+	VK_OPERATION_DEFERRED_KHR = 1000268002,
+	VK_OPERATION_NOT_DEFERRED_KHR = 1000268003,
+	VK_ERROR_INVALID_VIDEO_STD_PARAMETERS_KHR = -1000299000,
+	VK_ERROR_COMPRESSION_EXHAUSTED_EXT = -1000338000,
+	VK_PIPELINE_BINARY_MISSING_KHR = 1000483000,
+	VK_ERROR_NOT_ENOUGH_SPACE_KHR = -1000483000,
+	VK_ERROR_INCOMPATIBLE_SHADER_BINARY_EXT = 1000482000,
+	VK_RESULT_MAX_ENUM = 2147483647,
+};
+
+pub fn checkResult(result: c.VkResult) void {
+	const resultEnum = std.meta.intToEnum(VkResultEnum, result) catch {
+		std.log.err("Encountered a vulkan error with unknown error code {}", .{result});
+		return;
+	};
+	if(resultEnum == .VK_SUCCESS) return;
+	std.log.err("Encountered a vulkan error: {s}", .{@tagName(resultEnum)});
+}
+
+fn checkResultIfAvailable(result: anytype) void {
+	if(@TypeOf(result) != void) {
+		checkResult(result);
+	}
+}
+
+fn allocEnumerationGeneric(function: anytype, allocator: NeverFailingAllocator, args: anytype) []@typeInfo(@typeInfo(@TypeOf(function)).@"fn".params[@typeInfo(@TypeOf(function)).@"fn".params.len - 1].type.?).pointer.child {
+	const T = @typeInfo(@typeInfo(@TypeOf(function)).@"fn".params[@typeInfo(@TypeOf(function)).@"fn".params.len - 1].type.?).pointer.child;
+	var count: u32 = 0;
+	checkResultIfAvailable(@call(.auto, function, args ++ .{&count, null}));
+	const list = allocator.alloc(T, count);
+	checkResultIfAvailable(@call(.auto, function, args ++ .{&count, list.ptr}));
+	return list;
+}
+
+pub fn enumerateInstanceLayerProperties(allocator: NeverFailingAllocator) []c.VkLayerProperties {
+	return allocEnumerationGeneric(c.vkEnumerateInstanceLayerProperties, allocator, .{});
```

Comment:
As you probably know [vkEnumerateInstanceLayerProperties](https://registry.khronos.org/vulkan/specs/latest/man/html/vkEnumerateInstanceLayerProperties.html) is allowed to return `VK_INCOMPLETE` as a success status:

>  at most pPropertyCount structures will be written, and VK_INCOMPLETE will be returned instead of VK_SUCCESS, to indicate that not all the available properties were returned.

Reply:
This only happens if "pPropertyCount is less than the number of layer properties available", which is not possible since we allocate the buffer based on the size that Vulkan gives us.

Reply:
Hmm, I see, how do you suggest we resolve this? Should I just retry until the counts are the same?

Reply:
The solution I had in mind was getting rid of that logic since we require only one layer. However a valid solution would be to allow for pPropertyCount or less (if VK_INCOMPLETE was returned) and, if we don't find our required layers there, to again query pPropertyCount with `vkEnumerateInstanceLayerProperties` ensure new layers were not added. All of that is a lot of code just for sake of correctness, so if we didn't remove this code we might as well leave it as it is now.